### PR TITLE
WL-3170 Refactor and test email sending

### PR DESCRIPTION
**Pull request by ghost**
This was a total mess with a mix of different classes for different types of email, yet different behaviours for these emails being defined in a big switch statement.

There is still a lot of type specific behaviour defined outside of the individual email classes but this change specifically extracts the behaviour for generating a list of events for making iCal files. Common behaviours are shared between groups of emails by subclassing.

This also introduces the `isCancellation` method to each email class, which is used to determine whether the resulting calendar attachment should take advantage of the `CANCEL` method.

There are tests defined for each of the types of email for which we are generating events.

### CODE REVIEW

File: impl/src/test/org/sakaiproject/signup/logic/messages/CancellationEmailTest.java [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+    @Mock protected SakaiFacade _mockedFacade;
+    @Mock protected SignupCalendarHelper _mockedCalendarHelper;
+
+    @Mock protected SignupTimeslot _mockedTimeslot;
+    @Mock protected SignupTimeslot _mockedCancelledTimeslot;
+
+    @Before
+    public void setUp() {
+
+        MockitoAnnotations.initMocks(this);
```
Comment (buckett) [inline impl/src/test/org/sakaiproject/signup/logic/messages/CancellationEmailTest.java:44]: Not a problem, but you can use `@RunWith(MockitoJUnit44Runner.class)` on the class to be a bit more declarative.

File: impl/src/test/org/sakaiproject/signup/logic/messages/CancellationEmailTest.java [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+
+/**
+ * @author Ben Holmes
+ */
+public class CancellationEmailTest {
+
+    protected SignupEmailBase _email;
+
+    @Mock protected User _mockedUser;
+    @Mock protected SignupMeeting _mockedMeeting;
```
Comment (buckett) [inline impl/src/test/org/sakaiproject/signup/logic/messages/CancellationEmailTest.java:32]: Your ruby style showing through :grinning:

File: impl/src/test/org/sakaiproject/signup/logic/messages/CancellationEmailTest.java (hunk 1 above)
Comment (ghost) [inline impl/src/test/org/sakaiproject/signup/logic/messages/CancellationEmailTest.java:44]: Hey that's nice! Cool. :+1:

File: impl/src/java/org/sakaiproject/signup/logic/messages/OrganizerEmailBase.java [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+import org.sakaiproject.signup.model.SignupTimeslot;
+import org.sakaiproject.user.api.User;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+
+
+/**
+ * An email that is sent to an attendee when they are signed up
```
Comment (buckett) [inline impl/src/java/org/sakaiproject/signup/logic/messages/OrganizerEmailBase.java]: Shouldn't this be organizer?

**Comment by buckett** [comment]
I'm a little confused by the type hierarchy:

screen shot 2014-01-29 at 11 16 33

Some of the are based on the recipients (eg `OrganizerEmailBase` and `AttendeeEmailBase`) and some of them are based on the the action `TransferEmailBase`. Does `TransferEmailBase` not extend a recipient type because the same email is sent to both Organizers and Attendees?

**Comment by buckett** [comment]
From you work on this do you know why the emailing classes don't use the signup User object but instead depend directly on the sakai one?

**Comment by ghost** [comment]
I have just tried to come up with names to describe the common behaviour, I agree they're bad choices and if anything I'm making the non-descriptive class names worse. I also don't like that those two cancellation emails have their own behaviour, and so don't have anything to inherit.

Perhaps a sensible choice would be to rename `AttendeeEmailBase` to `NewSignupEmailBase`, since all of these events are triggered by an attendee signing up to a timeslot. That said, the `AttendeeSignupEmail` is also triggered by a signup, even though its behaviour is to email the organiser.

The `OrganizerEmailBase` is concerned with emails that generate a full ical event of attendees for the overall meeting.

**Comment by ghost** [comment]
Maybe the classes could focus on what type of events they generate:
- `AllUsersEmailBase` generates meeting or timeslot events depending on whether or not the recipient is an organiser.
- `OrganizerEmailBase` generates a meeting event
-  `AttendeeEmailBase` generates timeslot event(s)
- `TransferEmailBase` generates two timeslot events (one is cancelled)
- `AttendeeCancellationOwnEmail` and `CancellationEmail` generate one cancelled timeslot

**Comment by ghost** [comment]
I have no idea why the plain Sakai `User` is used here but I will look into it after lunch.

**Comment by buckett** [comment]
Don't worry about the Sakai `User` issue, it was just if you knew.

File: impl/src/java/org/sakaiproject/signup/logic/messages/AllUsersEmailBase.java [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+abstract public class AllUsersEmailBase extends SignupEmailBase {
+
+    /**
+     * {@inheritDoc}
+     */
+    public List<ExtEvent> generateEvents(User user, SignupCalendarHelper calendarHelper) {
+
+        List<ExtEvent> events = new ArrayList<ExtEvent>();
+
+        if (this.userIsAnOrganiser(user)) {
```
Comment (buckett) [inline impl/src/java/org/sakaiproject/signup/logic/messages/AllUsersEmailBase.java:25]: Generally in Java people tend not to prefix method calls with this. Typically this is only used to disambiguate between method arguments and instance variable. Again, doesn't need fixing, just something that "feels" different.

Comment (ghost) [inline impl/src/java/org/sakaiproject/signup/logic/messages/AllUsersEmailBase.java:25]: Got it.

File: impl/src/java/org/sakaiproject/signup/logic/messages/OrganizerEmailBase.java (hunk 3 above)
Comment (ghost) [inline impl/src/java/org/sakaiproject/signup/logic/messages/OrganizerEmailBase.java]: Corrected in 66d411b

**Comment by buckett** [comment]
I like the mocking and the verification of cancellation.

**Comment by ghost** [comment]
Cheers

**Pull request merged**